index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/EsIndex.java

```diff
@@ -0,0 +1,308 @@
+     * @param workspace workspace name where this index will be created.
+     */
+    public EsIndex(EsClient client, ExecutionContext context, IndexDefinition defn, String workspace) {
+        this.client = client;
+        this.name = defn.getName();
+        this.workspace = workspace;
+        this.columns = new EsIndexColumns(context, defn);
+        this.operations = new Operations(context.getValueFactories(), columns);
+        this.createIndex();
+    }
+
+    /**
+     * Creates new index.
+     * 
+     * @param client provides access to the elasticsearch functions.
+     * @param columns columns definition.
+     * @param context Modeshape execution context.
+     * @param name the name of the index
+     * @param workspace workspace name where this index will be created.
+     */
+    protected EsIndex(EsClient client, EsIndexColumns columns, ExecutionContext context, String name, String workspace) {
+        this.client = client;
+        this.name = name;
+        this.workspace = workspace;
+        this.columns = columns;
+        this.operations = new Operations(context.getValueFactories(), columns);
+        this.createIndex();
+    }
+
+    /**
+     * Generates local index name.
+     * 
+     * @return 
+     */
+    private String name() {
+        return name.toLowerCase() + "-" + workspace;
+    }
+
+    /**
+     * Executes create index action.
+     */
+    private void createIndex() {
+        try {
+            client.createIndex(name(), workspace, columns.mappings(workspace));
+            client.flush(name());
+        } catch (IOException e) {
+            throw new EsIndexException(e);
+        }
+    }
+
+    @Override
+    public void add(String nodeKey, String propertyName, Object value) {
+        CheckArg.isNotNull(nodeKey, "nodeKey");
+        CheckArg.isNotNull(propertyName, "propertyName");
+        CheckArg.isNotNull(value, "value");
+
+        EsIndexColumn column = columns.column(propertyName);
+        assert column != null : "Unexpected column for the index " + name();
+
+        try {
+            EsDocument doc = findOrCreateDoc(nodeKey);
+            putValue(doc, column, value);
+            client.storeDocument(name(), workspace, nodeKey, doc);
+            client.refresh(name());
+        } catch (IOException e) {
+            throw new EsIndexException(e);
+        }
+    }
+
+    @Override
+    public void add(String nodeKey, String propertyName, Object[] values) {
+        CheckArg.isNotNull(nodeKey, "nodeKey");
+        CheckArg.isNotNull(propertyName, "propertyName");
+        CheckArg.isNotNull(values, "values");
+
+        EsIndexColumn column = columns.column(propertyName);
+        assert column != null : "Unexpected column for the index " + name();
+
+        try {
+            EsDocument doc = findOrCreateDoc(nodeKey);
+            putValues(doc, column, values);
+            client.storeDocument(name(), workspace, nodeKey, doc);
+            client.refresh(name());
```

Comment:
@okulikov, can you explain why `client.refresh(...)` is called after every add and remove rather than during the commit of changes to the index? Perhaps I don't understand the API well enough, but ES refresh seems to cause a refresh of the index so that any changes since the last refresh are immediately visible.